packages/backend/src/services/warehouseClients/DatabricksWarehouseClient.ts

```diff
@@ -0,0 +1,187 @@
+import odbc, { Result } from 'odbc';
+import { CreateDatabricksCredentials, DimensionType } from 'common';
+import { WarehouseCatalog, WarehouseClient } from '../../types';
+import {
+    ParseError,
+    WarehouseConnectionError,
+    WarehouseQueryError,
+} from '../../errors';
+
+export const DRIVER_PATH = '/opt/simba/spark/lib/64/libsparkodbc_sb64.so';
+
+type SparkSchemaResult = {
+    TABLE_CAT: string;
+    TABLE_SCHEM: string;
+    TABLE_NAME: string;
+    COLUMN_NAME: string;
+    DATA_TYPE: number;
+    TYPE_NAME: string;
+};
+
+enum DatabricksTypes {
+    BOOLEAN = 'BOOLEAN',
+    BYTE = 'BYTE',
+    TINYINT = 'TINYINT',
+    SHORT = 'SHORT',
+    SMALLINT = 'SMALLINT',
+    INT = 'INT',
+    INTEGER = 'INTEGER',
+    LONG = 'LONG',
+    BIGINT = 'BIGINT',
+    FLOAT = 'FLOAT',
+    REAL = 'REAL',
+    DOUBLE = 'DOUBLE',
+    DATE = 'DATE',
+    TIMESTAMP = 'TIMESTAMP',
+    STRING = 'STRING',
+    BINARY = 'BINARY',
+    DECIMAL = 'DECIMAL',
+    DEC = 'DEC',
+    NUMERIC = 'NUMERIC',
+    INTERVAL = 'INTERVAL', // INTERVAL HOUR
+    ARRAY = 'ARRAY', // ARRAY<type>
+    STRUCT = 'STRUCT', // STRUCT<type,type...>
+    MAP = 'MAP',
+    CHAR = 'CHAR',
+    VARCHAR = 'VARCHAR',
+}
+
+const normaliseDatabricksType = (type: string): string => {
+    const r = /^[A-Z]+/;
+    const match = r.exec(type);
+    if (match === null) {
+        throw new ParseError(
+            `Cannot understand type from Databricks: ${type}`,
+            {},
+        );
+    }
+    return match[0];
+};
```

Comment:
Could `type.toUpperCase()` replace this function ?

Reply:
This is because spark has some really specific types and I strip off extra info:
* `DECIMAL(12,9)` -> becomes `DECIMAL`
* `ARRAY<BIGINT>` -> becomes `ARRAY`
* `INTERVAL HOUR` -> becomes `INTERVAL`

Reply:
The regex is to greedily match all uppercase chars starting from the beginning of the string

Reply:
Ah good to know :) all good then